packages/studio-base/src/hooks/useIndexedDbRecents.ts

```diff
@@ -0,0 +1,163 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/
+
+import { set as idbSet, get as idbGet, createStore as idbCreateStore } from "idb-keyval";
+import { isEqual } from "lodash";
+import { useCallback, useLayoutEffect, useMemo, useRef, useState } from "react";
+import { useAsync } from "react-use";
+import { v4 as uuid } from "uuid";
+
+import Logger from "@foxglove/log";
+
+const log = Logger.getLogger(__filename);
+
+const IDB_KEY = "recents";
+
+type RecentRecordCommon = {
+  // Record id - use IndexedDbRecentsStore.GenerateRecordId() to generate
+  id: string;
+
+  // The source id
+  sourceId: string;
+
+  // The primary text for the recent record
+  title: string;
+
+  // Optional label for the recent record
+  label?: string;
+};
+
+type RecentConnectionRecord = RecentRecordCommon & {
+  type: "connection";
+  // Optional arguments stored with the recent entry
+  extra?: Record<string, string | undefined>;
+};
+
+type RecentFileRecord = RecentRecordCommon & {
+  type: "file";
+  handle: FileSystemFileHandle;
+};
+
+type UnsavedRecentRecord = Omit<RecentConnectionRecord, "id"> | Omit<RecentFileRecord, "id">;
+
+type RecentRecord = RecentConnectionRecord | RecentFileRecord;
+
+interface IRecentsStore {
+  // Recent records
+  recents: RecentRecord[];
+
+  // Add a new recent
+  addRecent: (newRecent: UnsavedRecentRecord) => void;
+
+  // Save changes
+  save: () => Promise<void>;
+}
+
+function useIndexedDbRecents(): IRecentsStore {
+  const idbStore = useMemo(() => idbCreateStore("foxglove-recents", "recents"), []);
+
+  const { value: initialRecents } = useAsync(async () => {
+    const untypedRecents = await idbGet(IDB_KEY, idbStore);
+    return untypedRecents as RecentRecord[];
+  }, [idbStore]);
```

Comment:
You could make this a bit more concise right?

```suggestion
  const { value: initialRecents } = useAsync(
    async () => await idbGet<RecentRecord[]>(IDB_KEY, idbStore),
    [idbStore],
  );
```